server/src/main/java/keywhiz/service/resources/admin/SecretsResource.java

```diff
@@ -696,6 +697,74 @@ public Response updateCurrentSecretVersion(@Auth User user, @PathParam("secretId
     return Response.noContent().build();
   }
 
+  /**
+   * Clones a secret, creating a duplicate of all its data under a new name. Group assignments are not copied; callers
+   * should handle that separately.
+   *
+   * @param user     the admin user performing this operation
+   * @param request  a request object containing the current secret name and the new name to use for the clone
+   * @return 200 with new secret details if secret was cloned, 404 if not found
+   * <p>
+   * description Clones a secret to a new name. Used by Keywhiz CLI.
+   * <p>
+   * responseMessage 200 Cloned the secret to a new name
+   * <p>
+   * responseMessage 400 Secret cannot be created with the new name
+   * <p>
+   * responseMessage 404 Secret with original name not found
+   * <p>
+   * responseMessage 409 A secret with the new name already exists
+   */
+  @Path("clone")
+  @Timed @ExceptionMetered
+  @POST
+  public Response cloneSecret(@Auth User user, @Valid CloneSecretRequestV2 request) {
+    logger.info("User '{}' cloned secret original name={}, clone name={}", user, request.name(), request.newName());
+    Secret existingSecret = secretController.getSecretByName(request.name()).orElseThrow(NotFoundException::new);
+    long newId;
+    try {
+      Secret newSecret = secretController.builder(existingSecret.getName(),
+              existingSecret.getSecret(),
+              existingSecret.getCreatedBy(),
+              existingSecret.getExpiry())
+              .withDescription(existingSecret.getDescription())
+              .withMetadata(existingSecret.getMetadata())
+              .withOwnerName(existingSecret.getOwner())
+              .withType(existingSecret.getType().orElse(""))
+              .create();
+      newId = newSecret.getId();
+    } catch (DataAccessException e) {
```

Comment:
What about the secret contents? I'm worried that we'll have a new top-level secret with no actual contents/content history pointing to it. Or is this handled magically by the SecretController logic? Or do we not actually want the whole history cloned?

Reply:
Actually you're right, the content history is lost. I had another implementation that used the SecretDAO directly and I think we need to do that + add a DAO method to clone the actual history.